scripts/run_portserver.py

```diff
@@ -0,0 +1,425 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Adapted from portserver.py in the python_portpicker project (commit
+# bf6250b121cd610c3280cd748b5702a2be040f41) at
+# https://github.com/google/python_portpicker/blob/main/src/portserver.py,
+# which is Copyright 2015 Google Inc. All Rights Reserved. The following
+# changes have been made by The Oppia Authors to this file since it was
+# copied from the python_portpicker project:
+#
+# * We used threading instead of asyncio for compatibility with Python 2.
+#   This change involved adding a Server class to replace asyncio's
+#   event loop.
+# * We re-wrote the logger format string to be compatible with Python 2.
+#   We kept the same general string structure.
+# * We renamed _handle_port_request to handle_port_request and removed
+#   the original handle_port_request function from the
+#   _PortServerRequestHandler class. We also made _handle_port_request
+#   return the request response instead of writing it to a provided
+#   writer.
+# * We re-wrote the main function to use the Server class instead of
+#   asyncio. In doing so we also simplified the main function by
+#   removing the signal.SIGUSR1 handler.
+# * We changed the default portserver address to `portserver.sock` to
+#   avoid problems introduced by using an address beginning with a null
+#   byte.
+
+"""A server to hand out network ports to applications running on one host.
+Typical usage:
+ 1) Run one instance of this process on each of your unittest farm hosts.
+ 2) Set the PORTSERVER_ADDRESS environment variable in your test runner
+    environment to let the portpicker library know to use a port server
+    rather than attempt to find ports on its own.
+$ /path/to/portserver.py &
+$ export PORTSERVER_ADDRESS=portserver.sock
+$ # ... launch a bunch of tests that use portpicker ...
+"""
+
+import argparse
+import collections
+import logging
+import os
+import socket
+import sys
+import threading
+
+log = None  # Initialized to a logging.Logger by _configure_logging().
+
+_PROTOS = [(socket.SOCK_STREAM, socket.IPPROTO_TCP),
+           (socket.SOCK_DGRAM, socket.IPPROTO_UDP)]
+
+
+def _get_process_command_line(pid):
+    try:
+        with open('/proc/{}/cmdline'.format(pid), 'rt') as cmdline_f:
+            return cmdline_f.read()
+    except IOError:
+        return ''
+
+
+def _get_process_start_time(pid):
+    try:
+        with open('/proc/{}/stat'.format(pid), 'rt') as pid_stat_f:
+            return int(pid_stat_f.readline().split()[21])
+    except IOError:
+        return 0
+
+
+def _bind(port, socket_type, socket_proto):
+    """Try to bind to a socket of the specified type, protocol, and port.
+    For the port to be considered available, the kernel must support at least
+    one of (IPv6, IPv4), and the port must be available on each supported
+    family.
+    Args:
```

Comment:
Try to clean all this up so that it follows our existing docstring style; add typeinfo for args, etc.

Reply:
Done!